crypto/sig_abstractions.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package crypto
+
+import "fmt"
+
+// AlgorithmType enum type for signing algorithms
+type AlgorithmType uint64
+
+// all AlgorithmType enums
+const (
+	minAlgorithmType AlgorithmType = iota
+
+	PlaceHolderType
+
+	maxAlgorithmType
+)
+
+func (t AlgorithmType) isValidType() bool {
+	return minAlgorithmType < t && t < maxAlgorithmType
+}
+
+// ByteSignature is a cryptographic signature represented by bytes.
+type ByteSignature []byte
+
+// Signer interface represents the possible operations that can be done with a signing key.
+// outputs Sign, SignBytes which are self explanatory and GetVerifier which is a representation of a public key.
+type Signer interface {
+	Sign(message Hashable) ByteSignature
+	SignBytes(message []byte) ByteSignature
+	GetVerifyingKey() VerifyingKey
+}
+
+// ErrBadSignature represents a bad signature
+var ErrBadSignature = fmt.Errorf("invalid signature")
+
+// Verifier interface represent a public key of a SignatureAlgorithm.
+// Verifier returns error for bad signature/ other issues while verifying a signature, or nil for correct signature -
+// that is, returns: complain or no complain.
+type Verifier interface {
+	Verify(message Hashable, sig ByteSignature) error
+	VerifyBytes(message []byte, sig ByteSignature) error
+}
+
+// SignatureAlgorithm holds a Signer, and the type of algorithm the Signer conforms to.
+// to add a key - verify that PackedSignatureAlgorithm's function (getSigner) returns your key.
+type SignatureAlgorithm struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Type AlgorithmType            `codec:"sigType"`
+	Pack PackedSignatureAlgorithm `codec:"keys"`
+}
+
+// VerifyingKey is the correct way to interact with a Verifier. It implements the interface,
```

Comment:
VerifyingKey is the abstraction of the verifying key. it can return the correct key according to the underlying algorithm